sdk/containerregistry/azure-containers-containerregistry/src/main/java/com/azure/containers/containerregistry/specialized/ContainerRegistryBlobAsyncClient.java

```diff
@@ -322,34 +317,40 @@ public Mono<DownloadBlobResult> downloadBlob(String digest) {
      * @throws NullPointerException thrown if the {@code digest} is null.
      */
     @ServiceMethod(returns = ReturnType.SINGLE)
-    public Mono<Response<DownloadBlobResult>> downloadBlobWithResponse(String digest) {
-        return withContext(context -> this.downloadBlobWithResponse(digest, context));
+    public Mono<BlobDownloadAsyncResult> downloadStream(String digest) {
+        return withContext(context -> downloadBlobInternal(digest, context));
     }
 
-    private Mono<Response<DownloadBlobResult>> downloadBlobWithResponse(String digest, Context context) {
+    private Mono<BlobDownloadAsyncResult> downloadBlobInternal(String digest, Context context) {
         if (digest == null) {
             return monoError(LOGGER, new NullPointerException("'digest' can't be null."));
         }
 
-        return this.blobsImpl.getBlobWithResponseAsync(repositoryName, digest, context).flatMap(streamResponse -> {
-            String resDigest = UtilsImpl.getDigestFromHeader(streamResponse.getHeaders());
-
-            BinaryData binaryData = streamResponse.getValue();
-
-            // The service wants us to validate the digest here since a lot of customers forget to do it before consuming
-            // the contents returned by the service.
-            if (Objects.equals(resDigest, digest)) {
-                Response<DownloadBlobResult> response = new SimpleResponse<>(
-                    streamResponse.getRequest(),
-                    streamResponse.getStatusCode(),
-                    streamResponse.getHeaders(),
-                    new DownloadBlobResult(resDigest, binaryData));
-
-                return Mono.just(response);
-            } else {
-                return monoError(LOGGER, new ServiceResponseException("The digest in the response does not match the expected digest."));
-            }
-        }).onErrorMap(UtilsImpl::mapException);
+        Flux<ByteBuffer> content =
+            blobsImpl.getChunkWithResponseAsync(repositoryName, digest, new HttpRange(0, (long) CHUNK_SIZE).toString(), context)
+                .flatMapMany(firstResponse -> getAllChunks(firstResponse, digest, context))
+                .flatMapSequential(chunk -> chunk.getValue().toFluxByteBuffer(), 1);
+        return Mono.just(ConstructorAccessors.createBlobDownloadResult(digest, content));
+    }
+
+    private Flux<ContainerRegistryBlobsGetChunkResponse> getAllChunks(ContainerRegistryBlobsGetChunkResponse firstResponse, String digest, Context context) {
```

Comment:
I'll leave this as an open question. 

This design is a bit different from `azure-storage-blob` where chunked download is only leveraged when performing download to file operations as disk writing should support non-sequential writes. Or, better put if we concurrently download ranges 0-4MB and 12MB-16MB, and before 4-8MB and 8MB-12MB complete, this is fine as the file ranges can still be written correctly.

This however is a case of we're downloading an unknown size container, which in this case `azure-storage-blob` uses a concept of open-ended download, `Range: start-` which indicates download from a given start point until completion. The download is then tracked for progression, such as 12MB has been downloaded, and resuming logic is added on top if there is a failure, so if at 12MB a network error happens another download call is made with `Range: progress-` to resume the download until either completion or exhaustion of retries (such as a complete networking failure).

To me this should be following the latter logic.

Reply:
yes, thanks for pointing this out and explaining how it works. I'd like to do it after finalizing the feature work and API changes.  I created #33382 to track it.